checker/src/main/java/org/checkerframework/checker/calledmethods/CalledMethodsVisitor.java

```diff
@@ -0,0 +1,85 @@
+import org.checkerframework.checker.calledmethods.qual.CalledMethods;
+import org.checkerframework.checker.compilermsgs.qual.CompilerMessageKey;
+import org.checkerframework.common.accumulation.AccumulationVisitor;
+import org.checkerframework.common.basetype.BaseTypeChecker;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.javacutil.TreeUtils;
+
+/**
+ * This visitor implements the custom error message finalizer.invocation.invalid, and supports
+ * counting the number of framework build calls.
+ */
+public class CalledMethodsVisitor extends AccumulationVisitor {
+
+    /**
+     * Custom error message key for finalizer invocations. The custom error message makes it easier
+     * for users to tell what methods weren't called.
+     */
+    public static final @CompilerMessageKey String FINALIZER_INVOCATION_INVALID =
+            "finalizer.invocation.invalid";
+
+    /**
+     * Constructor matching super.
+     *
+     * @param checker the type-checker associated with this visitor
+     */
+    public CalledMethodsVisitor(final BaseTypeChecker checker) {
+        super(checker);
+    }
+
+    @Override
+    public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
+
+        if (checker.getBooleanOption(CalledMethodsChecker.COUNT_FRAMEWORK_BUILD_CALLS)) {
+            ExecutableElement element = TreeUtils.elementFromUse(node);
+            for (FrameworkSupport frameworkSupport :
+                    ((CalledMethodsAnnotatedTypeFactory) getTypeFactory()).getFrameworkSupports()) {
+                if (frameworkSupport.isBuilderBuildMethod(element)) {
+                    ((CalledMethodsChecker) checker).numBuildCalls++;
+                    break;
+                }
+            }
+        }
+        return super.visitMethodInvocation(node, p);
+    }
+
+    /**
+     * Adds special reporting for method.invocation.invalid errors to turn them into
+     * finalizer.invocation.invalid errors.
+     */
+    @Override
+    protected void reportMethodInvocabilityError(
+            MethodInvocationTree node, AnnotatedTypeMirror found, AnnotatedTypeMirror expected) {
+
+        AnnotationMirror expectedCM = expected.getAnnotation(CalledMethods.class);
+        if (expectedCM != null) {
+            AnnotationMirror foundCM = found.getAnnotation(CalledMethods.class);
+            Set<String> foundMethods =
+                    foundCM == null
+                            ? Collections.emptySet()
+                            : new HashSet<>(atypeFactory.getAccumulatedValues(foundCM));
+            List<String> expectedMethods = atypeFactory.getAccumulatedValues(expectedCM);
+            StringBuilder missingMethods = new StringBuilder();
+            for (String expectedMethod : expectedMethods) {
+                if (!foundMethods.contains(expectedMethod)) {
+                    missingMethods.append(expectedMethod);
+                    missingMethods.append("() ");
```

Comment:
Rather than having a trailing space at the end of missingMethods, use a StringJoiner.